_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+### Database as backbone for data integration
+
+As we said, databases are experts on their own domain. Data integration, with different sources and sinks, lands outside of that domain. We'd like to use the same code that we use for operational transformations (with the same testing and deployment processes) to integrate data. Let's see how we can solve this issue using application code.
+
+### Dual writes for data integration
+
+With this approach, the application is now in charge of updating every single data store. This sounds like a good idea but it introduces a couple of issues.
+
+  * Race conditions: if two clients send updates onto the same logical record, there is no guarantee that those updates will be applied ordered into the different datastores. That means that the app needs to implement itself Isolation from ACID properties, and that's not a trivial endeavour.
+  * Coupling: suddenly, what it was a simple app that writes to a database, it's a monster with coupling into different datastores. This approach doesn't scale organisationally as the owners of that service will have to serve requests for every single team.
+  * Crash recovery: even in the absence of concurrency this approach has an issue with Atomicity from ACID properties. What happens if the first updates succeed but the last one fails? It's not feasible to provide distributed heterogenous transactions, so, again, app will have to provide a do-it-yourself implementation of transactions.
+
+### Reflections on derived state
+
+Before of reaching to the proposed solution, let's think deeply about the concept of SSOT and derived data. Why do we keep different views of the original data? Let's see the reasoning behind different examples:
+
+  * Materialised views: our SSOT will have a format that won't satisfy every single query pattern. Maybe it's a normalised relational database as we need to save on storage costs and we were looking for strong referential constraints. We might have a multiple joins query that takes a lot of time, so we decided to persist that view on the database, to speed up reads. Database will take care of updating the materialised view as soon as the original tables gets updated.
+
+  * Indexes: every table, column family or whatever data store abstraction provides an efficient way of accessing its data. Maybe that's not enough for our query patterns, so we'll need to provide an index. Again, database will be in charge of updating it.
+
+  * Caches: some part of our data will be accessed more often. Also, there are physical storages that are faster, and more expensive on the other side. We use caches to keep that 'popular' data into efficient and spare data storage.
```

Comment:
You're right. It can be rephrased to "Also, there are physical storages that are faster and more expensive"